### PR TITLE
fix: local storage reservations fixes

### PR DESCRIPTION
**Pull request by magik6k**
## Proposed Changes
More debugging of accounting issues in the storage reservation mechanism in Curio.

* This PR adds a lot more debug logging to storage reservation code
* Fixes a bug in curio task storage where in some cases we'd release storage multiple times
  * Fixed with sync.once as other ways of fixing it would be either very complicated or would trade robustness in favour of sometimes maybe not releasing storage
* Also a small drive-by fix which drops existing SDR layers on sdr restart - seems like this was one of the causes of corrupted sectors

Currently testing on my mainnet setup, seems to work very well so far

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: storage/paths/local.go [hunk 1]
```diff
@@ -438,7 +468,24 @@ func (st *Local) Reserve(ctx context.Context, sid storiface.SectorRef, ft storif
...
+	done := func() {
+		if doneCalled {
+			curStack := make([]byte, 20480)
+			curStackN := runtime.Stack(curStack, false)
+
+			log.Errorw("double done call", "sector", sid, "fileType", ft, "prevStack", string(firstDonebuf[:firstDoneN]), "curStack", string(curStack[:curStackN]))
+		}
+
+		firstDonebuf = make([]byte, 20480)
+		firstDoneN = runtime.Stack(firstDonebuf, false)
```
Comment (snadrus) [inline storage/paths/local.go:484]: firstDonebuf = firstDonebuf[:runtime.Stack(firstDonebuf, false)]
Then you can get rid of firstDoneN

File: storage/paths/local.go [hunk 2]
```diff
@@ -438,7 +468,24 @@ func (st *Local) Reserve(ctx context.Context, sid storiface.SectorRef, ft storif
...
+
+		doneCalled = true
+	}
+
 	deferredDone := func() { done() }
 	defer func() {
 		st.localLk.Unlock()
```
Comment (snadrus) [inline storage/paths/local.go:491]: Lets get rid of the confusing done(): 
1. Name Reserve()'s returns: (release func(), err error)
2. Here, if err != nil { log_the_stack } // is this the intent of the first done() ? 
3. The L527 done() could be inline. 
A double-done should be impossible.

**Review by snadrus — CHANGES_REQUESTED** [review]

File: storage/paths/local.go (hunk 2 above)
Comment (magik6k) [reply storage/paths/local.go:491]: > Here, if err != nil { log_the_stack } // is this the intent of the first done() ?

No, the intent of this stack logging was identifying where this double call was happening.

Can drop that code, but decided to leave as it doesn't do much harm, and if we ever hit that again the visibility will be very handy.

Comment (magik6k) [reply storage/paths/local.go:490]: > The L527 done() could be inline.

Can you elaborate what do you mean by that? That code just appends some cleanup logic to the `done` call chain; The call chain will only ever be called once by the defer if Reserve fails at any later point, or if it succeeds, by the release return.

The bug here was that in the successful case the release function could be called more than once by the task_storage code, which was a bug in that code.

**Review by snadrus — APPROVED** [review]

**Review by snadrus — APPROVED** [review]

**Pull request merged**